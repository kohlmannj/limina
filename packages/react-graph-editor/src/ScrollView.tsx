import React, { Component, CSSProperties } from 'react';
import styled, { css } from 'react-emotion';
import { OverflowMode } from './index';
import ScrollBar from './ScrollBar';

export interface IScrollViewProps {
  className?: string;
  dangerouslySetInnerHTML?: {
    __html: string;
  };
  overflow?: OverflowMode;
  overflowX?: OverflowMode;
  overflowY?: OverflowMode;
  style?: CSSProperties;
}

export interface IScrollViewState {
  progressX: number;
  progressY: number;
  scaleX: number;
  scaleY: number;
}

const ScrollBarCorner = styled('div')`
  position: absolute;
  bottom: 0;
  right: 0;
  z-index: 9999;
  width: ${props => props.theme.trackWidth || 15}px;
  height: ${props => props.theme.trackWidth || 15}px;
  background: ${props => props.theme.trackColor || '#efefef'};
`;

const ScrollViewContainer = styled('div')`
  position: relative;
  overflow: hidden;
`;

const ScrollViewOverflowContainer = styled('div')`
  width: ${props =>
    `calc(100% - ${
      props.theme && typeof props.theme.trackWidth === 'number' ? props.theme.trackWidth : 15
    }px)`};
  height: ${props =>
    `calc(100% - ${
      props.theme && typeof props.theme.trackWidth === 'number' ? props.theme.trackWidth : 15
    }px)`};
  overflow: auto;

  ::-webkit-scrollbar {
    display: none;
  }
`;

export default class ScrollView extends Component<IScrollViewProps, IScrollViewState> {
  public state: IScrollViewState = {
    progressX: 0,
    progressY: 0,
    scaleX: 1,
    scaleY: 1,
  };

  private root?: HTMLDivElement | null;

  private content?: HTMLDivElement | null;

  public componentDidMount() {
    this.startListening();
  }

  public componentWillUnmount() {
    this.stopListening();
  }

  public render() {
    const {
      children,
      className,
      dangerouslySetInnerHTML,
      overflow,
      overflowX,
      overflowY,
      style,
      ...rest
    } = this.props;
    const { progressX, progressY, scaleX, scaleY } = this.state;

    return (
      <ScrollViewContainer
        innerRef={e => {
          this.root = e;
        }}
        className={className}
        style={style}
        {...rest}
      >
        <ScrollViewOverflowContainer
          innerRef={e => {
            this.content = e;
          }}
          dangerouslySetInnerHTML={dangerouslySetInnerHTML}
        >
          {children}
        </ScrollViewOverflowContainer>
        <ScrollBar
          className={css`
            position: absolute;
            z-index: 1;
          `}
          orientation="horizontal"
          overflow={overflowX || overflow}
          progress={progressX}
          scale={scaleX}
        />
        <ScrollBar
          className={css`
            position: absolute;
            z-index: 1;
          `}
          orientation="vertical"
          overflow={overflowY || overflow}
          progress={progressY}
          scale={scaleY}
        />
        <ScrollBarCorner />
      </ScrollViewContainer>
    );
  }

  private onScroll = (): void => {
    if (!this.content) {
      return;
    }

    const { width, height } = this.content.getBoundingClientRect();
    const { scrollTop, scrollHeight, scrollLeft, scrollWidth } = this.content;

    this.setState({
      progressX: scrollLeft / (scrollWidth - width),
      progressY: scrollTop / (scrollHeight - height),
      scaleX: Math.max(scrollWidth / Math.round(width), 1),
      scaleY: Math.max(scrollHeight / Math.round(height), 1),
    });
  };

  private startListening = () => {
    if (this.content) {
      this.content.addEventListener('scroll', this.onScroll);
      this.content.addEventListener('resize', this.onScroll);
      this.onScroll();
    }
  };

  private stopListening = () => {
    if (this.content) {
      this.content.removeEventListener('scroll', this.onScroll);
      this.content.removeEventListener('resize', this.onScroll);
    }
  };
}
