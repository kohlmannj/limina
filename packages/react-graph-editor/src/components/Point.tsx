/** @jsx jsx */
import { css, jsx, SerializedStyles } from '@emotion/core';
import { FunctionComponent } from 'react';
import styled, { StyledComponent } from '@emotion/styled';
import { ViewBoxContext } from './ViewBoxContext';
import { ThemeProps, ViewBox, OriginDirection, PointShape } from '..';

export interface PointProps {
  className?: string;
  selected?: boolean;
  x: number;
  y: number;
}

export interface StyledButtonProps extends PointProps, ViewBox {}

const getOriginTranslateFactor = (origin: OriginDirection) =>
  origin === 'left' || origin === 'top' ? -0.5 : 0.5;

export type DynamicStyle = (props: PointProps) => SerializedStyles | undefined;

export const shapes: Record<PointShape, DynamicStyle> = {
  circle: () => css`
    border-radius: 50%;
  `,
  diamond: () => undefined,
  square: () => undefined,
};

const selectedClassName = (props: PointProps & { theme?: ThemeProps }) => css`
  border-color: ${props.theme && props.theme.color ? props.theme.color : '#000'};
  background: #fff;
  box-shadow: 0 0 5px 2px rgba(59, 153, 252, 0.75);
`;

const translate = (props: StyledButtonProps) =>
  props.originX && props.originY
    ? `translate(${getOriginTranslateFactor(props.originX) * 100}%, ${getOriginTranslateFactor(
        props.originY
      ) * 100}%)`
    : '';

const position = (props: StyledButtonProps & { theme?: ThemeProps }) => css`
  position: absolute;
  ${props.originX && `${props.originX}: ${((props.x - props.minX) / props.width) * 100}%;`}
  ${props.originY && `${props.originY}: ${((props.y - props.minY) / props.height) * 100}%;`}
  transform: ${translate(props)} ${
  props.theme && props.theme.shape === 'diamond' ? 'rotate(45deg)' : ''
};
`;

// eslint-disable-next-line @typescript-eslint/no-unused-vars
const Button: FunctionComponent<StyledButtonProps> = ({ className, selected, x, y, ...rest }) => (
  <button
    type="button"
    className={className}
    aria-label={`Point (${x}, ${y})`}
    title={`(${x}, ${y})`}
    {...rest}
  />
);

const StyledButton: StyledComponent<StyledButtonProps, StyledButtonProps, ThemeProps> = styled(
  Button
)`
  width: ${props => `${props.theme.thickness || 8}px`};
  height: ${props => `${props.theme.thickness || 8}px`};
  box-sizing: border-box;
  padding: 0;
  border: 1px solid transparent;
  background: ${props => props.theme.color || '#000'};
  box-shadow: 0 0 1px 1px rgba(0, 0, 0, 0.15);
  cursor: move;
  font-size: 0;
  line-height: 0;
  ${props => (!props.theme || props.theme.shape === 'circle') && `border-radius: 50%`};
  ${props => props.selected && selectedClassName};
  ${position};
  outline: 0;
  text-indent: 100%;
`;

export const Point: FunctionComponent<PointProps> = props => (
  <ViewBoxContext.Consumer>
    {(viewBox: ViewBox) => <StyledButton {...props} {...viewBox} />}
  </ViewBoxContext.Consumer>
);

Point.defaultProps = {
  selected: false,
};
