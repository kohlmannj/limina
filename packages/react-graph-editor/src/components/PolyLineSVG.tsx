import React, { FunctionComponent } from 'react';
import styled, { StyledComponent } from '@emotion/styled';
import { getDimensionsForPoints } from '../utils';
import { ThemeProps, PointTuple } from '..';

export type InterpolationKeyword = 'linear' | 'ease-in-out' | 'ease-in' | 'ease-out';

export type InterpolationFunction = (x: number) => number;

export type InterpolationKeywordOrFunction = InterpolationKeyword | InterpolationFunction;

export interface PolyLineSVGProps {
  className?: string;
  interpolation?: InterpolationKeywordOrFunction;
  label?: string;
  points: PointTuple[];
}

export const PolyLineSVG: FunctionComponent<PolyLineSVGProps> = ({
  className,
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  label,
  points,
  ...rest
}) => (
  <svg xmlns="http://www.w3.org/2000/svg" {...getDimensionsForPoints(points)}>
    <polyline
      className={className}
      vectorEffect="non-scaling-stroke"
      points={points.map(point => point.join(',')).join(' ')}
      {...rest}
    />
  </svg>
);

export const StyledPolyLineSVG: StyledComponent<
  PolyLineSVGProps,
  PolyLineSVGProps,
  ThemeProps
> = styled(PolyLineSVG)`
  fill: none;
  stroke: ${props => props.theme.color || '#000'};
  stroke-width: ${props => `${props.theme.thickness / 4 || 2}px`};
`;
