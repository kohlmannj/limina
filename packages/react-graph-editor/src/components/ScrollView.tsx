import React, { Component, CSSProperties } from 'react';
import styled, { css } from 'react-emotion';
import { OverflowMode } from '..';
import { defaultTheme } from '../theme';
import ScrollBar from './ScrollBar';
import ScrollBarCorner from './ScrollBarCorner';

export interface IScrollViewProps {
  className?: string;
  dangerouslySetInnerHTML?: {
    __html: string;
  };
  overflow?: OverflowMode;
  overflowX?: OverflowMode;
  overflowY?: OverflowMode;
  style?: CSSProperties;
}

export interface IScrollViewState {
  progressX: number;
  progressY: number;
  scaleX: number;
  scaleY: number;
}

const ScrollViewContainer = styled.div`
  position: relative;
  overflow: hidden;
`;

const ScrollViewOverflowContainer = styled.div`
  overflow: auto;
  width: ${props =>
    `calc(100% - ${
      props.theme && typeof props.theme.trackWidth !== 'undefined'
        ? props.theme.trackWidth
        : defaultTheme.trackWidth
    }px)`};
  height: ${props =>
    `calc(100% - ${
      props.theme && typeof props.theme.trackWidth !== 'undefined'
        ? props.theme.trackWidth
        : defaultTheme.trackWidth
    }px)`};

  ::-webkit-scrollbar {
    display: none;
  }
`;

export default class ScrollView extends Component<IScrollViewProps, IScrollViewState> {
  public state: IScrollViewState = {
    progressX: 0,
    progressY: 0,
    scaleX: 1,
    scaleY: 1,
  };

  private root?: HTMLDivElement | null;

  private content?: HTMLDivElement | null;

  public componentDidMount() {
    this.startListening();
  }

  public componentWillUnmount() {
    this.stopListening();
  }

  public render() {
    const {
      children,
      className,
      dangerouslySetInnerHTML,
      overflow,
      overflowX,
      overflowY,
      style,
      ...rest
    } = this.props;
    const { progressX, progressY, scaleX, scaleY } = this.state;

    return (
      <ScrollViewContainer
        innerRef={e => {
          this.root = e;
        }}
        className={className}
        style={style}
        {...rest}
      >
        <ScrollViewOverflowContainer
          innerRef={e => {
            this.content = e;
          }}
          dangerouslySetInnerHTML={dangerouslySetInnerHTML}
        >
          {children}
        </ScrollViewOverflowContainer>
        <ScrollBar
          className={css`
            position: absolute;
            z-index: 1;
          `}
          orientation="horizontal"
          overflow={overflowX || overflow}
          progress={progressX}
          scale={scaleX}
        />
        <ScrollBar
          className={css`
            position: absolute;
            z-index: 1;
          `}
          orientation="vertical"
          overflow={overflowY || overflow}
          progress={progressY}
          scale={scaleY}
        />
        <ScrollBarCorner />
      </ScrollViewContainer>
    );
  }

  private onScroll = (): void => {
    if (!this.content) {
      return;
    }

    const { width, height } = this.content.getBoundingClientRect();
    const { scrollTop, scrollHeight, scrollLeft, scrollWidth } = this.content;

    this.setState({
      progressX: scrollLeft / (scrollWidth - width),
      progressY: scrollTop / (scrollHeight - height),
      scaleX: Math.max(scrollWidth / Math.round(width), 1),
      scaleY: Math.max(scrollHeight / Math.round(height), 1),
    });
  };

  private startListening = () => {
    if (this.content) {
      this.content.addEventListener('scroll', this.onScroll);
      this.content.addEventListener('resize', this.onScroll);
      this.onScroll();
    }
  };

  private stopListening = () => {
    if (this.content) {
      this.content.removeEventListener('scroll', this.onScroll);
      this.content.removeEventListener('resize', this.onScroll);
    }
  };
}
