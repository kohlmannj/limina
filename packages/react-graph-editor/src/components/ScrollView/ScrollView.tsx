/** @jsx jsx */
/* eslint-env browser */
/* eslint-disable react/no-danger */
import { jsx } from '@emotion/core';
import { Component, CSSProperties } from 'react';
import { ScrollBar } from '../ScrollBar';
import { ScrollBarCorner } from '../ScrollBarCorner';
import { absolutelyPositioned, overflowContainer, root } from './styles';
import { getNormalizedDragEventData } from './utils';
import { OverflowMode, ScrollBarAxis } from '../../index';
import {
  StyledThumbSegmentProps,
  ThumbSegmentPosition,
} from '../ScrollBar/components/ThumbSegment';

export type DragType = 'progress' | 'scale';

export interface ScrollViewProps {
  className?: string;
  contentClassName?: string;
  contentStyle?: CSSProperties;
  dangerouslySetInnerHTML?: {
    __html: string;
  };
  maxScaleX?: number;
  maxScaleY?: number;
  minScaleX?: number;
  minScaleY?: number;
  overflow?: OverflowMode;
  overflowX?: OverflowMode;
  overflowY?: OverflowMode;
  proportional?: boolean;
  scaleIncrement?: number;
  style?: CSSProperties;
  thumbXEndProps?: Partial<StyledThumbSegmentProps>;
  thumbXMiddleProps?: Partial<StyledThumbSegmentProps>;
  thumbXStartProps?: Partial<StyledThumbSegmentProps>;
  thumbYEndProps?: Partial<StyledThumbSegmentProps>;
  thumbYMiddleProps?: Partial<StyledThumbSegmentProps>;
  thumbYStartProps?: Partial<StyledThumbSegmentProps>;
}

export interface ScrollViewState {
  dragAxis?: ScrollBarAxis;
  dragBeginX?: number;
  dragBeginY?: number;
  dragPosition?: ThumbSegmentPosition;
  dragStartProgress?: number;
  dragThumbLength?: number;
  dragTrackLength?: number;
  dragType?: DragType;
  progressX: number;
  progressY: number;
  scaleX: number;
  scaleY: number;
}

export interface ScaleEventHandlerOptions {
  axis: ScrollBarAxis;
  pos: ThumbSegmentPosition;
  type: DragType;
}

export type ReactMouseOrTouchEvent<E> = React.MouseEvent<E> | React.TouchEvent<E>;

export class ScrollView extends Component<ScrollViewProps, ScrollViewState> {
  public static defaultProps = {
    maxScaleX: 10,
    maxScaleY: 10,
    minScaleX: 1,
    minScaleY: 1,
    scaleIncrement: 0.01,
    scaleX: 1,
    scaleY: 1,
  };

  public state: ScrollViewState = {
    progressX: 0.5,
    progressY: 0.5,
    scaleX: 1,
    scaleY: 1,
  };

  private overflowContainer?: HTMLDivElement | null;

  public componentDidMount() {
    if (typeof window !== 'undefined') {
      this.startListeningToWindowEvents();
    }
  }

  public componentDidUpdate(_prevProps: ScrollViewProps, prevState: ScrollViewState) {
    const { dragAxis, dragType, progressX, progressY, scaleX, scaleY } = this.state;

    if (!this.overflowContainer) {
      return;
    }

    // While dragging: recalculate scrollLeft and scrollTop positions after scale changes
    if (dragType) {
      const { width, height } = this.overflowContainer.getBoundingClientRect();

      if (
        dragAxis === 'x' &&
        (scaleX !== prevState.scaleX || progressX !== prevState.progressX) &&
        progressX > 0 &&
        progressX < 1
      ) {
        const nextScrollLeft = (this.overflowContainer.scrollWidth - width) * progressX;
        this.overflowContainer.scrollLeft = nextScrollLeft;
      } else if (
        dragAxis === 'y' &&
        (scaleY !== prevState.scaleY || progressY !== prevState.progressY) &&
        progressY > 0 &&
        progressY < 1
      ) {
        const nextScrollTop = (this.overflowContainer.scrollHeight - height) * progressY;
        this.overflowContainer.scrollTop = nextScrollTop;
      }
    }
  }

  public componentWillUnmount() {
    this.stopListeningToWindowEvents();
  }

  private onScroll = (): void => {
    const { dragType } = this.state;
    if (!this.overflowContainer || typeof dragType !== 'undefined') {
      return;
    }

    const { width, height } = this.overflowContainer.getBoundingClientRect();
    const { scrollTop, scrollHeight, scrollLeft, scrollWidth } = this.overflowContainer;

    const progressX = scrollLeft / (scrollWidth - width);
    const progressY = scrollTop / (scrollHeight - height);

    if (!Number.isNaN(progressX)) {
      this.setState({ progressX });
    }

    if (!Number.isNaN(progressY)) {
      this.setState({ progressY });
    }

    this.setState({
      scaleX: Math.max(scrollWidth / Math.round(width), 1),
      scaleY: Math.max(scrollHeight / Math.round(height), 1),
    });
  };

  private startListeningToWindowEvents = () => {
    window.addEventListener('resize', this.onScroll);

    if (this.overflowContainer) {
      this.onScroll();
    }
  };

  private stopListeningToWindowEvents = () => {
    const { dragType } = this.state;

    window.removeEventListener('resize', this.onScroll);

    if (typeof dragType !== 'undefined') {
      window.removeEventListener('mousemove', this.onMoveDrag);
      window.removeEventListener('mouseup', this.onEndDrag);
      window.removeEventListener('touchmove', this.onMoveDrag);
      window.removeEventListener('touchend', this.onEndDrag);
    }
  };

  private onBeginDrag = ({
    axis: dragAxis,
    pos: dragPosition,
    type: dragType,
  }: ScaleEventHandlerOptions) => (e: ReactMouseOrTouchEvent<HTMLButtonElement>): void => {
    const { progressX, progressY } = this.state;

    const {
      clientX: dragBeginX,
      clientY: dragBeginY,
      thumbLength: dragThumbLength,
      touches,
      trackLength: dragTrackLength,
    } = getNormalizedDragEventData(e, dragAxis);

    let dragStartProgress;

    if (dragAxis === 'x') {
      dragStartProgress = progressX;
    } else if (dragAxis === 'y') {
      dragStartProgress = progressY;
    }

    this.setState({
      dragAxis,
      dragBeginX,
      dragBeginY,
      dragPosition,
      dragStartProgress,
      dragThumbLength,
      dragTrackLength,
      dragType,
    });

    window.addEventListener(touches ? 'touchmove' : 'mousemove', this.onMoveDrag);
    window.addEventListener(touches ? 'touchend' : 'mouseup', this.onEndDrag, {
      once: true,
    });
  };

  private onMoveDrag = (e: MouseEvent | TouchEvent): void => {
    const { minScaleX, maxScaleX, minScaleY, maxScaleY, scaleIncrement } = this.props;
    const {
      dragBeginX,
      dragBeginY,
      dragAxis,
      dragPosition,
      dragStartProgress,
      dragThumbLength,
      dragTrackLength,
      dragType,
      progressX,
      progressY,
      scaleX,
      scaleY,
    } = this.state;

    if (!dragBeginX || !dragBeginY || !dragAxis) {
      return;
    }

    const { clientX, clientY, trackLength } = getNormalizedDragEventData(e, dragAxis);

    const dX = clientX - dragBeginX;
    const dY = clientY - dragBeginY;

    // Figure out which axis we're constrained to
    let currentProgress;
    let currentScale;
    let dLength;
    let minScale;
    let maxScale;

    if (dragAxis === 'x') {
      currentProgress = progressX;
      currentScale = scaleX;
      dLength = dragPosition === 'start' ? -dX : dX;
      minScale = minScaleX;
      maxScale = maxScaleX;
    } else if (dragAxis === 'y') {
      currentProgress = progressY;
      currentScale = scaleY;
      dLength = dragPosition === 'start' ? -dY : dY;
      minScale = minScaleY;
      maxScale = maxScaleY;
    }

    // Calculate length scale factor
    let dLengthScaleFactor = 1;

    if (typeof currentProgress !== 'undefined' && currentProgress > 0 && currentProgress < 1) {
      if (dragPosition === 'start') {
        dLengthScaleFactor = 1 / currentProgress;
      } else if (dragPosition === 'end') {
        dLengthScaleFactor = 1 / (1 - currentProgress);
      }
    }

    // Scaling Mode
    if (
      dragType === 'scale' &&
      typeof dragThumbLength !== 'undefined' &&
      typeof dLength !== 'undefined' &&
      typeof dLengthScaleFactor !== 'undefined' &&
      typeof dragTrackLength !== 'undefined' &&
      typeof minScale !== 'undefined' &&
      typeof maxScale !== 'undefined'
    ) {
      const nextThumbLength = dragThumbLength + dLength * dLengthScaleFactor;

      let nextScale = dragTrackLength / nextThumbLength;
      // Round nextScale to the nearest increment
      if (typeof scaleIncrement !== 'undefined') {
        nextScale = Math.round((nextScale * 1) / scaleIncrement) / (1 / scaleIncrement);
      }

      if (nextScale >= minScale && nextScale <= maxScale) {
        if (dragAxis === 'x') {
          this.setState({ scaleX: nextScale });
        } else if (dragAxis === 'y') {
          this.setState({ scaleY: nextScale });
        }
      }
    }
    // Progress Mode
    else if (
      dragType === 'progress' &&
      typeof dLength !== 'undefined' &&
      typeof dragStartProgress !== 'undefined' &&
      typeof currentScale !== 'undefined' &&
      currentScale > 1
    ) {
      const rawNextProgress = (dragStartProgress * trackLength + dLength) / trackLength;
      const nextProgress = Math.min(Math.max(rawNextProgress || 0, 0), 1);

      if (dragAxis === 'x') {
        this.setState({ progressX: nextProgress });
      } else if (dragAxis === 'y') {
        this.setState({ progressY: nextProgress });
      }
    }
  };

  private onEndDrag = (e: MouseEvent | TouchEvent): void => {
    const { touches } = e as TouchEvent;
    const { dragType, scaleX, scaleY } = this.state;

    if (dragType === 'scale') {
      if (scaleX === 1) {
        this.setState({ progressX: 0.5 });
      }

      if (scaleY === 1) {
        this.setState({ progressY: 0.5 });
      }
    }

    window.removeEventListener(touches ? 'touchmove' : 'mousemove', this.onMoveDrag);

    this.setState({
      dragAxis: undefined,
      dragBeginX: undefined,
      dragBeginY: undefined,
      dragPosition: undefined,
      dragStartProgress: undefined,
      dragThumbLength: undefined,
      dragTrackLength: undefined,
      dragType: undefined,
    });
  };

  public render() {
    /* eslint-disable @typescript-eslint/no-unused-vars */
    const {
      children,
      className,
      contentClassName,
      contentStyle: contentStyleProp,
      dangerouslySetInnerHTML,
      maxScaleX,
      maxScaleY,
      minScaleX,
      minScaleY,
      overflow,
      overflowX,
      overflowY,
      proportional,
      scaleIncrement,
      style,
      thumbXStartProps,
      thumbXMiddleProps,
      thumbXEndProps,
      thumbYStartProps,
      thumbYMiddleProps,
      thumbYEndProps,
      ...rest
    } = this.props;
    /* eslint-enable @typescript-eslint/no-unused-vars */
    const { progressX, progressY, scaleX, scaleY } = this.state;

    const contentStyle: CSSProperties = {
      ...contentStyleProp,
      width: `${scaleX * 100}%`,
      height: `${scaleY * 100}%`,
    };

    // Generate event handlers
    const onBeginDragXSrt = this.onBeginDrag({ axis: 'x', pos: 'start', type: 'scale' });
    const onBeginDragXMid = this.onBeginDrag({ axis: 'x', pos: 'middle', type: 'progress' });
    const onBeginDragXEnd = this.onBeginDrag({ axis: 'x', pos: 'end', type: 'scale' });
    const onBeginDragYSrt = this.onBeginDrag({ axis: 'y', pos: 'start', type: 'scale' });
    const onBeginDragYMid = this.onBeginDrag({ axis: 'y', pos: 'middle', type: 'progress' });
    const onBeginDragYEnd = this.onBeginDrag({ axis: 'y', pos: 'end', type: 'scale' });

    return (
      <div className={[className, root].filter(Boolean).join(' ')} style={style} {...rest}>
        <div
          css={overflowContainer({})}
          onScroll={this.onScroll}
          ref={e => {
            this.overflowContainer = e;
          }}
        >
          <div
            className={contentClassName}
            dangerouslySetInnerHTML={dangerouslySetInnerHTML}
            style={contentStyle}
          >
            {children}
          </div>
        </div>
        <ScrollBar
          axis="x"
          css={absolutelyPositioned}
          overflow={overflowX || overflow}
          progress={progressX}
          scale={scaleX}
          thumbStartProps={{
            dragSignifier: true,
            onMouseDown: onBeginDragXSrt,
            onTouchStart: onBeginDragXSrt,
          }}
          thumbMiddleProps={{
            onMouseDown: onBeginDragXMid,
            onTouchStart: onBeginDragXMid,
          }}
          thumbEndProps={{
            dragSignifier: true,
            onMouseDown: onBeginDragXEnd,
            onTouchStart: onBeginDragXEnd,
          }}
        />
        <ScrollBar
          axis="y"
          css={absolutelyPositioned}
          overflow={overflowY || overflow}
          progress={progressY}
          scale={scaleY}
          thumbStartProps={{
            dragSignifier: true,
            onMouseDown: onBeginDragYSrt,
            onTouchStart: onBeginDragYSrt,
          }}
          thumbMiddleProps={{
            onMouseDown: onBeginDragYMid,
            onTouchStart: onBeginDragYMid,
          }}
          thumbEndProps={{
            dragSignifier: true,
            onMouseDown: onBeginDragYEnd,
            onTouchStart: onBeginDragYEnd,
          }}
        />
        <ScrollBarCorner />
      </div>
    );
  }
}
