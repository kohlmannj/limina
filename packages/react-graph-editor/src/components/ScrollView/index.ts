export { default } from './ScrollView';
export * from './ScrollView';
export * from './ScrollView.d';
