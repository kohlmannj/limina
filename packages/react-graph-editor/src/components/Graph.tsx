/** @jsx jsx */
import { jsx } from '@emotion/core';
import chroma from 'chroma-js';
import { ThemeProvider } from 'emotion-theming';
import { CSSProperties, Fragment, FunctionComponent } from 'react';
import styled, { StyledComponent } from '@emotion/styled';
import { getViewBoxForPoints } from '../utils';
import { Point, shapes } from './Point';
import { PolyLineSVGProps } from './PolyLineSVG';
import { ViewBoxContext } from './ViewBoxContext';
import { ThemeProps, OriginDirection, PointTuple } from '..';

const shapeNames = Object.keys(shapes);

export interface GraphProps {
  className?: string;
  lines: PolyLineSVGProps[];
  minX?: number;
  minY?: number;
  maxX?: number;
  maxY?: number;
  originX?: OriginDirection;
  originY?: OriginDirection;
  style?: CSSProperties;
}

const Graph: FunctionComponent<GraphProps> = ({
  className,
  lines,
  minX: minXProp,
  minY: minYProp,
  maxX: maxXProp,
  maxY: maxYProp,
  originX,
  originY,
  ...rest
}) => {
  const autoViewBox = getViewBoxForPoints(
    lines.reduce((reducedPoints: PointTuple[], { points }) => [...reducedPoints, ...points], [])
  );

  const minX = typeof minXProp === 'number' ? minXProp : autoViewBox.minX;
  const minY = typeof minYProp === 'number' ? minYProp : autoViewBox.minY;
  const maxX = typeof maxXProp === 'number' ? maxXProp : autoViewBox.maxX;
  const maxY = typeof maxYProp === 'number' ? maxYProp : autoViewBox.maxY;

  const viewBox = {
    minX,
    minY,
    maxX,
    maxY,
    originX,
    originY,
    width: typeof minX === 'number' && typeof maxX === 'number' ? Math.abs(maxX - minX) : 0,
    height: typeof minY === 'number' && typeof maxY === 'number' ? Math.abs(maxY - minY) : 0,
  };

  return (
    <div className={className} {...rest}>
      <ViewBoxContext.Provider value={viewBox}>
        {lines.map(({ label, points }, index) => (
          <ThemeProvider
            key={label}
            theme={{
              color: chroma.brewer.Set1[index % chroma.brewer.Set1.length],
              shape: shapeNames[index % shapeNames.length],
            }}
          >
            <Fragment>
              {points.map(([x, y]) => (
                <Point key={`${label}_${x}_${y}`} x={x} y={y} />
              ))}
            </Fragment>
          </ThemeProvider>
        ))}
      </ViewBoxContext.Provider>
    </div>
  );
};

Graph.defaultProps = {
  originX: 'left',
  originY: 'bottom',
};

const StyledGraph: StyledComponent<GraphProps, GraphProps, ThemeProps> = styled(Graph)`
  position: relative;
  width: 100%;
  height: 100%;
`;

const GraphContainer: FunctionComponent<GraphProps> = ({ className, style, ...rest }) => (
  <div className={className} style={style}>
    <StyledGraph {...rest} />
  </div>
);

export const StyledGraphContainer = styled(GraphContainer)`
  width: 100%;
  height: 100%;
  box-sizing: border-box;
  padding: ${props => `${(props.theme.thickness || 8) / 2}px`};
`;
