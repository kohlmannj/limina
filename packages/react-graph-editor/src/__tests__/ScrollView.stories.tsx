/** @jsx jsx */
import { css, jsx } from '@emotion/core';
import { storiesOf } from '@storybook/react';
// import lipsum from '../__stubs__/lipsum';
import { ScrollView } from '../components/ScrollView';

storiesOf('react-graph-editor/ScrollView', module)
  // .add('with auto scale and y overflow', () => (
  //   <ScrollView
  //     style={{
  //       width: '50vw',
  //       height: '50vh',
  //       margin: '50px auto',
  //       background: 'gray',
  //     }}
  //     dangerouslySetInnerHTML={{ __html: lipsum }}
  //   />
  // ))
  // .add('with auto scale and x and y overflow', () => (
  //   <ScrollView
  //     style={{
  //       width: '50vw',
  //       height: '50vh',
  //       margin: '50px auto',
  //       background: 'gray',
  //     }}
  //   >
  //     <div style={{ width: '200%' }} dangerouslySetInnerHTML={{ __html: lipsum }} />
  //   </ScrollView>
  // ))
  // .add('with x and y overflow and custom theme', () => (
  //   <ThemeProvider
  //     theme={{
  //       trackClassName: css`
  //         background: navy;
  //       `,
  //       trackWidth: 20,
  //       thumbWidth: 20,
  //       thumbClassName: css`
  //         background: crimson;
  //         border-radius: 0;
  //       `,
  //     }}
  //   >
  //     <ScrollView
  //       style={{
  //         width: '50vw',
  //         height: '50vh',
  //         margin: '50px auto',
  //         background: 'gray',
  //       }}
  //       scaleX="auto"
  //       scaleY="auto"
  //     >
  //       <div style={{ width: '200%' }} dangerouslySetInnerHTML={{ __html: textContent }} />
  //     </ScrollView>
  //   </ThemeProvider>
  // ))
  .add('with resizeable kitten', () => (
    <ScrollView style={{ width: '100%', height: '100vh' }}>
      <div
        css={css`
          width: 100%;
          height: 100%;
          background-image: url('https://placekitten.com/1600/1600');
          background-size: 100% 100%;
        `}
      />
    </ScrollView>
  ));
