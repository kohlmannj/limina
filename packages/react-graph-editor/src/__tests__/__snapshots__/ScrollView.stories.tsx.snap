// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react-graph-editor/ScrollView with resizeable kitten 1`] = `
.emotion-4 {
  position: relative;
  overflow: hidden;
}

.emotion-2 {
  position: absolute;
  z-index: 1;
}

.emotion-0 {
  width: 100%;
  height: 100%;
  background-image: url('https://placekitten.com/g/1600/1600');
  background-size: 100% 100%;
}

.emotion-1 {
  overflow: auto;
  width: calc(100% - 16px);
  height: calc(100% - 16px);
  -webkit-overflow-scrolling: touch;
}

.emotion-1::-webkit-scrollbar {
  display: none;
}

<div
  className="emotion-4"
  scaleX={1}
  scaleY={1}
  style={
    Object {
      "height": "100vh",
      "width": "100%",
    }
  }
>
  <div
    className="emotion-1"
    onScroll={[Function]}
  >
    <div
      style={
        Object {
          "height": "100%",
          "width": "100%",
        }
      }
    >
      <div
        className="emotion-0"
      />
    </div>
  </div>
  <Component
    axis="x"
    className="emotion-2"
    progress={0.5}
    scale={1}
    thumbEndProps={
      Object {
        "dragSignifier": true,
        "onMouseDown": [Function],
        "onTouchStart": [Function],
      }
    }
    thumbMiddleProps={
      Object {
        "onMouseDown": [Function],
        "onTouchStart": [Function],
      }
    }
    thumbStartProps={
      Object {
        "dragSignifier": true,
        "onMouseDown": [Function],
        "onTouchStart": [Function],
      }
    }
  />
  <Component
    axis="y"
    className="emotion-2"
    progress={0.5}
    scale={1}
    thumbEndProps={
      Object {
        "dragSignifier": true,
        "onMouseDown": [Function],
        "onTouchStart": [Function],
      }
    }
    thumbMiddleProps={
      Object {
        "onMouseDown": [Function],
        "onTouchStart": [Function],
      }
    }
    thumbStartProps={
      Object {
        "dragSignifier": true,
        "onMouseDown": [Function],
        "onTouchStart": [Function],
      }
    }
  />
  <Styled(div) />
</div>
`;
